compiler/rustc_parse/src/parser/item.rs

```diff
@@ -562,6 +562,14 @@ impl<'a> Parser<'a> {
             self.sess.gated_spans.gate(sym::const_trait_impl, span);
         }
 
+        // Parse stray `impl async Trait`
+        if (self.token.span.at_least_rust_2018() && self.token.is_keyword(kw::Async))
+            || (self.token.span.is_rust_2015() && self.is_kw_followed_by_ident(kw::Async))
```

Comment:
Is it possible possible that the `span` of an `Interpolated` ident differs from the span of the corresponding `uninterpolate`'d ident? In other words does this PR break the following code?

```
// dep.rs  :::  rustc +stage1 dep.rs --crate-type=lib --edition=2021
#[macro_export]
macro_rules! gen { ($Trait:ident) => { impl $Trait for () {} } }

// usr.rs  :::  rustc +stage1 usr.rs --edition=2015 --extern=dep -L.
trait async {}
dep::gen! { async }
fn main() {}
```

If so, we should check `may_recover` first.

Reply:
Just checked out this branch and it does error:

<details><summary>summary</summary>

```
error: `async` trait implementations are unsupported
 --> usr.rs:3:1
  |
3 | dep::generate!(async);
  | ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `dep::generate` (in Nightly builds, run with -Z macro-backtrace for more info)

error: missing trait in a trait impl
 --> usr.rs:3:1
  |
3 | dep::generate!(async);
  | ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `dep::generate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: add a trait here
 --> /home/fmease/programming/rust2/dep.rs:2:56
  |
2 | macro_rules! generate { ($Trait:ident) => { impl $Trait Trait for () {} } }
  |                                                         +++++
help: for an inherent impl, drop this `for`
 --> /home/fmease/programming/rust2/dep.rs:2:56

2 - macro_rules! generate { ($Trait:ident) => { impl $Trait for () {} } }
2 + macro_rules! generate { ($Trait:ident) => { impl $Trait () {} } }
  |

warning: trait `async` should have an upper camel case name
 --> usr.rs:1:7
  |
1 | trait async {}
  |       ^^^^^ help: convert the identifier to upper camel case: `Async`
  |
  = note: `#[warn(non_camel_case_types)]` on by default

error: aborting due to 2 previous errors; 1 warning emitted
```

</details>

And `uninterpolate`'ing does indeed fix this issue but it requires inlining the definition of `is_kw_followed_by_ident`.